### ISSUE TITLE
📡 Implement Tensor Broadcasting Support

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#12]
# 📡 Implement Tensor Broadcasting Support

## 📋 Description

Currently, micrograd only supports element-wise operations between tensors of the same shape. We need to implement NumPy-style broadcasting to enable operations between tensors of different but compatible shapes.

## 🎯 Acceptance Criteria

- [ ] Implement broadcasting rules following NumPy semantics
- [ ] Support scalar-to-tensor broadcasting
- [ ] Support tensor-to-tensor broadcasting with compatible shapes
- [ ] Add `broadcast_tensors()` utility function
- [ ] Update all binary operations to support broadcasting
- [ ] Add comprehensive broadcasting tests
- [ ] Handle edge cases (empty tensors, 1D broadcasts)
- [ ] Optimize memory usage during broadcasting

### Broadcasting Rules (NumPy-compatible)
1. Start from trailing dimensions and work forward
2. Dimensions are compatible if:
   - They are equal, or
   - One of them is 1, or
   - One of them is missing
3. Missing dimensions are assumed to be 1

### Examples to Support
```python
# Scalar broadcasting
a = Tensor([[1, 2, 3], [4, 5, 6]])  # (2, 3)
b = Tensor(10)                       # ()
result = a + b                       # (2, 3)

# Vector broadcasting
a = Tensor([[1, 2, 3], [4, 5, 6]])  # (2, 3)
b = Tensor([10, 20, 30])             # (3,)
result = a + b                       # (2, 3)

# Matrix broadcasting
a = Tensor([[[1, 2]], [[3, 4]]])     # (2, 1, 2)
b = Tensor([[10, 20, 30]])           # (1, 3)
result = a + b                       # (2, 3, 2)
```

### Implementation Strategy

**Phase 1: Core Broadcasting Logic**
- [ ] `_broadcast_shapes()` - Determine output shape
- [ ] `_expand_tensor()` - Create broadcasted view
- [ ] `_is_broadcastable()` - Check compatibility

**Phase 2: Integration**
- [ ] Update `_elementwise_op()` in ops.py
- [ ] Modify autograd operations to handle broadcasting
- [ ] Update gradient computation for broadcasted operations

**Phase 3: Optimization**
- [ ] Memory-efficient broadcasting (avoid copying when possible)
- [ ] Stride-based broadcasting for views
- [ ] Cache broadcast shapes for repeated operations

## 🧪 Testing Requirements

- [ ] Test all broadcasting combinations from NumPy docs
- [ ] Test broadcasting with different dtypes
- [ ] Test memory efficiency (no unnecessary copies)
- [ ] Test gradient computation with broadcasting
- [ ] Performance benchmarks vs NumPy
- [ ] Edge cases: empty tensors, large dimension differences

## 📁 Files to Modify

- `grad/tensor.py` - Add broadcasting utilities
- `grad/autograd/ops.py` - Update elementwise operations
- `grad/utils/broadcasting.py` - New file for broadcasting logic
- `tests/test_broadcasting.py` - Comprehensive test suite

## 🔗 Dependencies

- Requires: Tensor stride system (✅ implemented)
- Requires: Shape manipulation (✅ implemented)
- Blocks: Matrix multiplication (#04)
- Blocks: Advanced tensor operations (#06)

### Memory Considerations
- Use stride manipulation instead of copying data when possible
- Implement lazy broadcasting (compute on access)
- Handle large tensor broadcasting efficiently

### Gradient Handling
- Broadcasted gradients need dimension reduction
- Sum gradients across broadcasted dimensions
- Preserve original tensor shapes in backward pass

### Error Handling
- Clear error messages for incompatible shapes
- Suggest compatible shapes when possible
- Validate broadcasting early to avoid computation

## 🏷️ Labels

- `enhancement`
- `core-functionality`
- `broadcasting`
- `high-priority`

## 📈 Priority

**High** - Required for most tensor operations and neural networks

## 🔗 Related Issues

- Blocks: Matrix multiplication (#04)
- Blocks: Loss functions (#09)
- Related to: Autograd backward pass (#01)
